core/src/main/scala/io/finch/request/package.scala

```diff
@@ -320,105 +371,33 @@ package object request extends LowPriorityImplicits {
+   * @param item the missing request item
    */
-  case class ParamNotFound(param: String) extends RequestReaderError(s"Param '$param' not found in the request.")
+  case class NotPresent(item: RequestItem) extends RequestReaderError(s"Required ${item.description} not present in the request.")
 
   /**
-   * An exception that indicates a broken validation rule on the param.
+   * An exception that indicates a broken validation rule on the request item.
    *
-   * @param param the param name
+   * @param item the invalid request item
    * @param rule the rule description
    */
-  case class ValidationFailed(param: String, rule: String)
-    extends RequestReaderError(s"Request validation failed: param '$param' $rule.")
-
-  /**
-   * An exception that indicates missed header in the request.
-   *
-   * @param header the missed header name
-   */
-  case class HeaderNotFound(header: String) extends RequestReaderError(s"Header '$header' not found in the request.")
-
+  case class NotValid(item: RequestItem, rule: String)
+    extends RequestReaderError(s"Validation failed: ${item.description} $rule.")
+  
   /**
-   * An exception that indicates a missing cookie in the request.
+   * An exception that indicates that a request item could be parsed.
    *
-   * @param cookie the missing cookie's name
+   * @param item the invalid request item
+   * @param targetType the type the item should be converted into
+   * @param cause the cause of the parsing error
    */
-  case class CookieNotFound(cookie: String) extends RequestReaderError(s"Cookie '$cookie' not found in the request.")
-
-  /**
-   * An exception that indicated a missing body in the request.
-   */
-  object BodyNotFound extends RequestReaderError("Body not found in the request.")
-
-  /**
-   * An exception that indicates an error in JSON format.
-   */
-  object BodyNotParsed extends RequestReaderError("A serialized object in a request body can not be parsed.")
-
-  /**
-   * An empty ''RequestReader''.
-   */
-  object EmptyReader extends RequestReader[Nothing] {
-    def apply[Req](req: Req)(implicit ev: Req => HttpRequest) =
-      new NoSuchElementException("Empty reader.").toFutureException
-  }
-
-  /**
-   * A const param.
-   */
-  object ConstReader {
-
-    /**
-     * Creates a ''RequestReader'' that reads given ''const'' param from the request.
-     *
-     * @return a const param value
-     */
-    def apply[A](const: Future[A]) = new RequestReader[A] {
-      def apply[Req](req: Req)(implicit ev: Req => HttpRequest) = const
-    }
-  }
-
-  private[this] object StringToValueOrFail {
-    def apply[A](param: String, rule: String)(number: => A) = new RequestReader[A] {
-      def apply[Req](req: Req)(implicit ev: Req => HttpRequest) =
-        try number.toFuture
-        catch { case _: IllegalArgumentException => ValidationFailed(param, rule).toFutureException }
+  case class NotParsed(item: RequestItem, targetType: ClassTag[_], cause: Throwable)
```

Comment:
How about to define an extra `upapply` method that just matches `item` to allow users write:

```scala
x match {
  case NotParsed(what) => ???
}
```